rewrite-java/src/main/java/org/openrewrite/java/internal/parser/TypeTable.java

```diff
@@ -0,0 +1,412 @@
+    /**
+     * Verifies that the bytecodes written out for the types represented in a type table
+     * will not be invalid and therefore rejected by the JVM verifier when used in a compilation
+     * step.
+     */
+    public static final String VERIFY_CLASS_WRITING = "org.openrewrite.java.TypeTableClassWritingVerification";
+
+    public static final String DEFAULT_RESOURCE_PATH = "META-INF/rewrite/classpath.tsv.snappy";
+
+    private static final Map<GroupArtifactVersion, Path> classesDirByArtifact = new LinkedHashMap<>();
+
+    public TypeTable(ExecutionContext ctx, Collection<String> artifactNames) {
+        this(ctx, findCaller().getClassLoader().getResourceAsStream(DEFAULT_RESOURCE_PATH), artifactNames);
+    }
+
+    public TypeTable(ExecutionContext ctx, InputStream is, Collection<String> artifactNames) {
+        try (InputStream snappy = new SnappyInputStream(is)) {
+            new Reader(ctx).read(snappy, artifactsNotYetWritten(artifactNames));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static Collection<String> artifactsNotYetWritten(Collection<String> artifactNames) {
+        Collection<String> notWritten = new ArrayList<>(artifactNames);
+        for (String artifactName : artifactNames) {
+            for (GroupArtifactVersion groupArtifactVersion : classesDirByArtifact.keySet()) {
+                if (Pattern.compile(artifactName + ".*")
+                        .matcher(groupArtifactVersion.getArtifactId() + "-" + groupArtifactVersion.getVersion())
+                        .matches()) {
+                    notWritten.remove(artifactName);
+                }
+            }
+        }
+        return notWritten;
+    }
+
+    /**
+     * Reads a type table from the classpath, and writes classes directories to disk for matching artifact names.
+     */
+    @RequiredArgsConstructor
+    static class Reader {
+        private final ExecutionContext ctx;
+        private GroupArtifactVersion gav;
+        private final Map<ClassDefinition, List<Member>> membersByClassName = new HashMap<>();
+
+        public void read(InputStream is, Collection<String> artifactNames) throws IOException {
+            if (is != null) {
+                try (BufferedReader in = new BufferedReader(new InputStreamReader(is))) {
+                    in.lines().skip(1).map(line -> line.split("\t", -1)).forEach(fields -> {
+                        GroupArtifactVersion rowGav = new GroupArtifactVersion(fields[0], fields[1], fields[2]);
+                        if (!rowGav.equals(gav)) {
+                            writeClassesDir();
+                        }
+
+                        String artifactVersion = fields[1] + "-" + fields[2];
+
+                        for (String artifactName : artifactNames) {
+                            if (Pattern.compile(artifactName + ".*").matcher(artifactVersion).matches()) {
```

Comment:
We could consider creating a map of precompiled patterns, so that we don't do that for every line in the file.

Reply:
Fixed in a23866b85faf9209880c1ebf17425b4c1feeda51.